feat: Integrate Surrealdb as VectorDB

dsaad68 opened the pull request:
# 🚀 SurrealDB Vector Database Integration

## 🔍 Overview
This PR adds support for [SurrealDB](https://surrealdb.com/) as a multi-model database backend for Agno. SurrealDB is a scalable, distributed database with native vector embedding support and HNSW indexes for vector search, making it a perfect fit for RAG applications.

### 🌟 About SurrealDB
SurrealDB is "the ultimate multi-model database for AI applications" that unifies vectors, graphs, documents, time-series and files in a single platform. 

### ✨ Features
- Added `SurrealVectorDb` implementation supporting both synchronous and asynchronous operations
- Implemented all standard vector database operations including:
  - Document insertion/upsert
  - Vector similarity search with filtering
  - Collection management (create, drop, exists)
- Added support for HNSW index configuration for performance tuning
- Added both sync and async demo files showing integration with PDFUrlKnowledgeBase

### 🛠️ Improvements
- 🔌 Extended the `AgentKnowledge` async capabilities with `async_filter_existing_documents`

### 📝 Usage Example
```python
from agno.vectordb.surrealdb import SurrealVectorDb

# Initialize SurrealDB vector database
surrealdb = SurrealVectorDb(
    url="ws://localhost:8000",
    username="root",
    password="root",
    namespace="test",
    database="test",
    collection="documents",
    efc=150,  # HNSW construction parameter
    m=12,     # HNSW connections parameter
    search_ef=40  # HNSW search parameter
)

# Use with any knowledge base implementation
knowledge_base = PDFUrlKnowledgeBase(
    urls=["https://example.com/document.pdf"],
    vector_db=surrealdb
)

# Load documents and search
knowledge_base.load(recreate=True)

# Create agent and query synchronously
agent = Agent(knowledge=knowledge_base, show_tool_calls=True, debug_mode=True)
agent.print_response("What are the 3 categories of Thai SELECT is given to restaurants overseas?", markdown=True)
```

## 🔗 References
- [SurrealDB Official Website](https://surrealdb.com/)
- [SurrealDB Documentation](https://surrealdb.com/docs)


## Type of change

- [ ] Bug fix
- [x] New feature
- [ ] Breaking change
- [x] Improvement
- [ ] Model update
- [ ] Other:

---

## Checklist

- [x] Code complies with style guidelines
- [x] Ran format/validation scripts (`./scripts/format.sh` and `./scripts/validate.sh`)
- [x] Self-review completed
- [x] Documentation updated (comments, docstrings)
- [x] Examples and guides: Relevant cookbook examples have been included or updated (if applicable)
- [x] Tested in clean environment
- [x] Tests added/updated (if applicable)

---

## Additional Notes

Add any important context (deployment instructions, screenshots, security considerations, etc.)

kausmeows commented on sandbox/demo_async_surrealdb.py:
```diff
@@ -0,0 +1,42 @@
+from agno.vectordb.surrealdb import SurrealVectorDb
+from agno.embedder.openai import OpenAIEmbedder
+
+# SurrealDB connection parameters
```
We keep all the example cookbooks in this path- `cookbook/agent_concepts/knowledge/vector_dbs`

dsaad68 replied:
@kausmeows Thanks a lot for your Feedback.

kausmeows replied:
got it, thanks!

dsaad68 replied:
Done!

kausmeows reviewed (commented):
Hey @dsaad68 thanks alot for this contribution. small comment

kausmeows commented on sandbox/qdrant-demo.py:
```diff
@@ -0,0 +1,29 @@
+
+from agno.agent import Agent
+from agno.knowledge.pdf_url import PDFUrlKnowledgeBase
+from agno.vectordb.qdrant import Qdrant
```
We already have cookbooks for qdrant, so this is not required

dsaad68 replied:
done!

kausmeows commented on sandbox/surrealdb-example.py:
```diff
@@ -0,0 +1,63 @@
+
+# Using a context manger to automatically connect and disconnect
+with Surreal("ws://localhost:8000/rpc") as db:
+    db.signin({"username": 'root', "password": 'root'})
```
This example is also not required. We only need two examples- `surreal_db.py` and `async_surreal_db.py`

dsaad68 replied:
As mentioned, I will do a cleanup before finalizing the PR.

dsaad68 replied:
Done!

kausmeows commented on sandbox/demo_surrealdb.py:
```diff
@@ -0,0 +1,44 @@
+from agno.agent import Agent
```
At the top of this cookbook can in a docstring can we mention the steps to setup surreal db before we can run this example?

dsaad68 replied:
Done!

dsaad68 commented:
@kausmeows could you please check if added `surrealdb` correctly to `pyproject.toml`?

kausmeows commented:
> @kausmeows could you please check if added `surrealdb` correctly to `pyproject.toml`?

Yes that looks fine but @dsaad68 i was testing this and ran the cookbook- `cookbook/agent_concepts/knowledge/vector_dbs/surrealdb/surrealdb.py`, but it did not work for me..? See below

![image](https://github.com/user-attachments/assets/9a9dca22-d031-4846-a69c-36295ebff0d7)

I ran the docker command mentioned, am i missing anything?

kausmeows commented on libs/agno/agno/vectordb/surrealdb/surrealdb.py:
```diff
@@ -0,0 +1,375 @@
+from agno.vectordb.distance import Distance
+
+
+class SurrealVectorDb(VectorDb):
```
sure np, let me know when i can test again

dsaad68 commented:
> > @kausmeows could you please check if added `surrealdb` correctly to `pyproject.toml`?
> 
> Yes that looks fine but @dsaad68 i was testing this and ran the cookbook- `cookbook/agent_concepts/knowledge/vector_dbs/surrealdb/surrealdb.py`, but it did not work for me..? See below
> 
> ![image](https://private-user-images.githubusercontent.com/82315953/445756181-9a9dca22-d031-4846-a69c-36295ebff0d7.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3NDc4NDQ5NjAsIm5iZiI6MTc0Nzg0NDY2MCwicGF0aCI6Ii84MjMxNTk1My80NDU3NTYxODEtOWE5ZGNhMjItZDAzMS00ODQ2LWE2OWMtMzYyOTVlYmZmMGQ3LnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNTA1MjElMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjUwNTIxVDE2MjQyMFomWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPTJkYTdhMmQyYzk3NGYzMDU5YWRlOTllNTNhMDQ2MjhjM2QzZTY5MGRhZmI1MWIxYmU2ZjNlNGNlZjAxZjNlYTgmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0In0.MqNVMF3RAcZtsq1wzGyIAjt2KSrUYMIaS8_zNH05tsI)
> 
> I ran the docker command mentioned, am i missing anything?

I have fixed this problem, the issue was with the version of the SurrealDB package. Can you run it again? 
Also, I have changed the name as you requested.

dsaad68 commented:
> > > > @kausmeows could you please check if added `surrealdb` correctly to `pyproject.toml`?
> > > 
> > > 
> > > Yes that looks fine but @dsaad68 i was testing this and ran the cookbook- `cookbook/agent_concepts/knowledge/vector_dbs/surrealdb/surrealdb.py`, but it did not work for me..? See below
> > > ![image](https://private-user-images.githubusercontent.com/82315953/445756181-9a9dca22-d031-4846-a69c-36295ebff0d7.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3NDc4NDQ5NjAsIm5iZiI6MTc0Nzg0NDY2MCwicGF0aCI6Ii84MjMxNTk1My80NDU3NTYxODEtOWE5ZGNhMjItZDAzMS00ODQ2LWE2OWMtMzYyOTVlYmZmMGQ3LnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNTA1MjElMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjUwNTIxVDE2MjQyMFomWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPTJkYTdhMmQyYzk3NGYzMDU5YWRlOTllNTNhMDQ2MjhjM2QzZTY5MGRhZmI1MWIxYmU2ZjNlNGNlZjAxZjNlYTgmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0In0.MqNVMF3RAcZtsq1wzGyIAjt2KSrUYMIaS8_zNH05tsI)
> > > I ran the docker command mentioned, am i missing anything?
> > 
> > 
> > I have fixed this problem, the issue was with the version of the SurrealDB package. Can you run it again? Also, I have changed the name as you requested.
> 
> @dsaad68 not sure if i'm doing something wrong but still getting the same error as in above screenshot... I also did `pip install -U surrealdb` to upgrade to the latest and also tried at version `1.0.3`

Sorry for the trouble. I will look into it.

However, a somewhat unconventional approach, could you please attempt to relocate the script to the other folder, such as the root directory, and then test it?

dsaad68 commented:
@kausmeowsI have fixed the problem. Now it should work.

martinschaer commented on libs/agno/pyproject.toml:
```diff
@@ -124,6 +124,7 @@ weaviate = ["weaviate-client"]
 milvusdb = ["pymilvus"]
 clickhouse = ["clickhouse-connect"]
 pinecone = ["pinecone==5.4.2"]
+surrealdb = ["surrealdb>=1.0.3"]
```
hello! 1.0.4 has been released. No functional or breaking changes, just improved types (specially the `query` return type) and some dependencies versions were relaxed for compatibility with other libraries

dsaad68 replied:
@martinschaer Could you please look at the queries?

martinschaer commented on libs/agno/agno/vectordb/surrealdb/surrealdb.py:
```diff
@@ -0,0 +1,397 @@
+        SELECT
+            content,
+            meta_data,
+            vector::distance::knn() as distance
```
is there a reasson why you use `vector::distance::knn() as distance` instead of `vector::similarity::cosine(vector, $vector) as similarity`, since the index was defined with `cosine`

dsaad68 commented on libs/agno/agno/vectordb/surrealdb/surrealdb.py:
```diff
@@ -0,0 +1,397 @@
+            content = $content,
+            embedding = $embedding,
+            meta_data = $meta_data;
+    """
```
@martinschaer I don't know where this query coming from 😅

dsaad68 replied:
Done!

dsaad68 commented on libs/agno/agno/vectordb/surrealdb/surrealdb.py:
```diff
@@ -0,0 +1,397 @@
+                if filters:
+                    data["meta_data"].update(filters)
+                log_debug(f"Upserting document: {doc.name} ({doc.meta_data})")
+                client.query(self.UPSERT_QUERY.format(collection=self.collection), data)
```
Done!

vinaydotblog commented:
@dsaad68 Is this still being worked upon, can I help with the tests ?

vinaydotblog commented:
@dsaad68 Thanks for your effort in building this, I'm really looking forward to this becoming part of the repo. I cloned your repo and ran ThaiRecipes.pdf test and it created separate table for each piece of the document. Is it the intended behaviour ?

## My Code
```
knowledge_base = PDFUrlKnowledgeBase(
    urls=["https://agno-public.s3.amazonaws.com/recipes/ThaiRecipes.pdf"],
    vector_db=SurrealDb(url="http://localhost:8072",
            namespace="dev",
            database="rag",
            username="root",
            password="dev@local"),
)
# Load the knowledge base: Comment out after first run
knowledge_base.load(recreate=True, upsert=True)
```

## The Result
![CleanShot 2025-07-09 at 12 43 50@2x](https://github.com/user-attachments/assets/d646aefe-af4c-4f7b-9085-c3b9071d6f2f)

dsaad68 commented:
> @dsaad68 Thanks for your effort in building this, I'm really looking forward to this becoming part of the repo. I cloned your repo and ran ThaiRecipes.pdf test and it created separate table for each piece of the document. Is it the intended behaviour ?
> 
> 
> 
> ## My Code
> 
> ```
> 
> knowledge_base = PDFUrlKnowledgeBase(
> 
>     urls=["https://agno-public.s3.amazonaws.com/recipes/ThaiRecipes.pdf"],
> 
>     vector_db=SurrealDb(url="http://localhost:8072",
> 
>             namespace="dev",
> 
>             database="rag",
> 
>             username="root",
> 
>             password="dev@local"),
> 
> )
> 
> # Load the knowledge base: Comment out after first run
> 
> knowledge_base.load(recreate=True, upsert=True)
> 
> ```
> 
> 
> 
> ## The Result
> 
> ![CleanShot 2025-07-09 at 12 43 50@2x](https://github.com/user-attachments/assets/d646aefe-af4c-4f7b-9085-c3b9071d6f2f)
> 
> 

No, it should not. Let me check everything. And get back to you.

martinschaer commented on libs/agno/agno/vectordb/surrealdb/surrealdb.py:
```diff
@@ -0,0 +1,392 @@
+            meta_data,
+            vector::distance::knn() as distance
+        FROM {collection}
+        WHERE embedding <|{limit},{distance}|> $query_embedding
```
Specifying here a `distance` function makes the query use brute force, instead of the HNSW index. The distance function should be set in the index instead.

```suggestion
        WHERE embedding <|{limit}|> $query_embedding
```

dsaad68 replied:
Thanks a lot for the clarifications.

martinschaer commented on libs/agno/agno/vectordb/surrealdb/surrealdb.py:
```diff
@@ -0,0 +1,397 @@
+        DEFINE FIELD IF NOT EXISTS content ON {collection} TYPE string;
+        DEFINE FIELD IF NOT EXISTS embedding ON {collection} TYPE array<float>;
+        DEFINE FIELD IF NOT EXISTS meta_data ON {collection} TYPE object;
+        DEFINE INDEX IF NOT EXISTS vector_idx ON {collection} FIELDS embedding HNSW DIMENSION {dimensions} DIST COSINE;
```
Set the distance function here instead of in the search query

```suggestion
        DEFINE INDEX IF NOT EXISTS vector_idx ON {collection} FIELDS embedding HNSW DIMENSION {dimensions} DIST {distance};
```

dsaad68 replied:
I have fixed it.

dsaad68 commented on libs/agno/agno/vectordb/surrealdb/surrealdb.py:
```diff
@@ -0,0 +1,392 @@
+                meta_data: Dict[str, Any] = doc.meta_data if isinstance(doc.meta_data, dict) else {}
+                data: Dict[str, Any] = {"content": doc.content, "embedding": doc.embedding, "meta_data": meta_data}
+                if filters:
+                    data["meta_data"].update(filters)
+                log_debug(f"Upserting document: {doc.name} ({doc.meta_data})")
+                client.query(self.UPSERT_QUERY.format(thing=doc.id if doc.id else self.collection), data)
```
Thanks a lot for the heads up. I fixed the both of them.

<img width="1048" height="574" alt="image" src="https://github.com/user-attachments/assets/27bb5e8b-f730-4a8b-b0fa-2f3d8e0ce7c5" />

dsaad68 commented:
> Thanks @dsaad68, but following error is still occuring.
> 
> ```
> INFO Embedder not provided, using OpenAIEmbedder as default.
> INFO Dropping collection
> Traceback (most recent call last):
>   File "/Users/vinayaggarwal/clear/play_with_ai_sdks/surreal_with_agno/agent.py", line 19, in <module>
>     knowledge_base.load(recreate=True, upsert=True)
>   File "/Users/vinayaggarwal/clear/play_with_ai_sdks/surreal_with_agno/.venv/lib/python3.12/site-packages/agno/knowledge/agent.py", line 108, in load
>     self.vector_db.drop()
>   File "/Users/vinayaggarwal/clear/play_with_ai_sdks/surreal_with_agno/.venv/lib/python3.12/site-packages/agno/vectordb/surrealdb/surrealdb.py", line 260, in drop
>     with self.connect() as client:
>          ^^^^^^^^^^^^^^
>   File "/usr/local/Cellar/python@3.12/3.12.9/Frameworks/Python.framework/Versions/3.12/lib/python3.12/contextlib.py", line 144, in __exit__
>     next(self.gen)
>   File "/Users/vinayaggarwal/clear/play_with_ai_sdks/surreal_with_agno/.venv/lib/python3.12/site-packages/agno/vectordb/surrealdb/surrealdb.py", line 151, in connect
>     self.sync_client.close()
>   File "/Users/vinayaggarwal/clear/play_with_ai_sdks/surreal_with_agno/.venv/lib/python3.12/site-packages/surrealdb/connections/sync_template.py", line 34, in close
>     raise NotImplementedError(f"close not implemented for: {self}")
> NotImplementedError: close not implemented for: <surrealdb.connections.blocking_http.BlockingHttpSurrealConnection object at 0x10e3a7140>
> ```

@vinaydotblog 

I’ve resolved the issue and successfully run both the synchronous and asynchronous examples using:
- `SURREALDB_URL = "ws://localhost:8000"`
- `SURREALDB_URL = "http://localhost:8000"`

Also, I have updated the tests.

dsaad68 commented:
> hey @dsaad68 really good work! Thanks again.
> 
> I noticed filtering is still not working. Could you check once please? <img alt="image" width="847" height="518" src="https://private-user-images.githubusercontent.com/82315953/466655926-81b41e58-a6c3-4a0c-a47c-f78f98ad385f.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3NTI2MDk2MTQsIm5iZiI6MTc1MjYwOTMxNCwicGF0aCI6Ii84MjMxNTk1My80NjY2NTU5MjYtODFiNDFlNTgtYTZjMy00YTBjLWE0N2MtZjc4Zjk4YWQzODVmLnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNTA3MTUlMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjUwNzE1VDE5NTUxNFomWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPWU2M2M1YzdmOTIzMmMxYmVjMWJmZmU3YmMzMjRhM2U3OGVmNDExOWI2MDE2ZmJhOTM1MmI5N2MxMjBlYWY4NWQmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0In0.ftnLth24Jnr__8mF5BcnMIBvrVgMqgHaxVDQJ0EpiYQ">
> 
> I ran this code (similar to what we have in: `cookbook/agent_concepts/knowledge/filters/pdf/filtering.py`)-
> 
> ```python
> from agno.agent import Agent
> from agno.knowledge.pdf import PDFKnowledgeBase
> from agno.utils.media import (
>     SampleDataFileExtension,
>     download_knowledge_filters_sample_data,
> )
> from agno.vectordb.surrealdb import SurrealDb
> 
> # SurrealDB connection parameters
> SURREALDB_URL = "ws://localhost:8000"
> SURREALDB_USER = "root"
> SURREALDB_PASSWORD = "root"
> SURREALDB_NAMESPACE = "test"
> SURREALDB_DATABASE = "test"
> 
> # Download all sample CVs and get their paths
> downloaded_cv_paths = download_knowledge_filters_sample_data(
>     num_files=5, file_extension=SampleDataFileExtension.PDF
> )
> 
> vector_db = SurrealDb(
>     url=SURREALDB_URL,
>     username=SURREALDB_USER,
>     password=SURREALDB_PASSWORD,
>     namespace=SURREALDB_NAMESPACE,
>     database=SURREALDB_DATABASE,
>     collection="recipes",  # Collection name for storing documents
>     efc=150,  # HNSW construction time/accuracy trade-off
>     m=12,  # HNSW max number of connections per element
>     search_ef=40,  # HNSW search time/accuracy trade-off
> )
> 
> 
> # Step 1: Initialize knowledge base with documents and metadata
> # ------------------------------------------------------------------------------
> # When initializing the knowledge base, we can attach metadata that will be used for filtering
> # This metadata can include user IDs, document types, dates, or any other attributes
> 
> knowledge_base = PDFKnowledgeBase(
>     path=[
>         {
>             "path": downloaded_cv_paths[0],
>             "metadata": {
>                 "user_id": "jordan_mitchell",
>                 "document_type": "cv",
>                 "year": 2025,
>             },
>         },
>         {
>             "path": downloaded_cv_paths[1],
>             "metadata": {
>                 "user_id": "taylor_brooks",
>                 "document_type": "cv",
>                 "year": 2025,
>             },
>         },
>         {
>             "path": downloaded_cv_paths[2],
>             "metadata": {
>                 "user_id": "morgan_lee",
>                 "document_type": "cv",
>                 "year": 2025,
>             },
>         },
>         {
>             "path": downloaded_cv_paths[3],
>             "metadata": {
>                 "user_id": "casey_jordan",
>                 "document_type": "cv",
>                 "year": 2025,
>             },
>         },
>         {
>             "path": downloaded_cv_paths[4],
>             "metadata": {
>                 "user_id": "alex_rivera",
>                 "document_type": "cv",
>                 "year": 2025,
>             },
>         },
>     ],
>     vector_db=vector_db,
> )
> 
> # Load all documents into the vector database
> knowledge_base.load(recreate=True)
> 
> # Step 2: Query the knowledge base with different filter combinations
> # ------------------------------------------------------------------------------
> 
> # Option 1: Filters on the Agent
> # Initialize the Agent with the knowledge base and filters
> agent = Agent(
>     knowledge=knowledge_base,
>     search_knowledge=True,
> )
> 
> agent.print_response(
>     "Tell me about Jordan Mitchell's experience and skills",
>     knowledge_filters={"user_id": "jordan_mitchell"},
>     markdown=True,
> )
> ```

I will look into it.

kausmeows commented on cookbook/agent_concepts/knowledge/vector_dbs/surrealdb/surreal_filter.py:
```diff
@@ -0,0 +1,105 @@
+from agno.agent import Agent
+from agno.knowledge.pdf import PDFKnowledgeBase
```
Thanks a lot for the push @dsaad68, can we please move this file to here- `cookbook/agent_concepts/knowledge/filters/filtering_surrealdb.py`

Then we good to merge!

dsaad68 commented on cookbook/agent_concepts/knowledge/vector_dbs/surrealdb/async_surreal.py:
```diff
@@ -0,0 +1,55 @@
+# Run SurrealDB in a container before running this script
+# docker run --rm --pull always -p 8000:8000 surrealdb/surrealdb:latest start --user root --pass root
+
```
@kausmeows 
I have run the same example, with this command:
```bash
uv run cookbook/agent_concepts/knowledge/vector_dbs/surrealdb/async_surreal_db.py
```

<img width="1219" height="755" alt="image" src="https://github.com/user-attachments/assets/846d72b7-f219-4db3-b8f9-14f0b0fe3b6d" />

kausmeows replied:
oh sorry, my bad. its all good!

kausmeows reviewed (approved):
Thanks a lot for this!

kausmeows commented:
@martinschaer appreciate all the help here 🚀

dsaad68 commented:
> Thanks a lot for this!

You are welcome! Happy to give back to the community!

vinaydotblog commented:
Congrats! @dsaad68 on getting this merged, I've also tested it in the main branch, it's working as aspected.

vinaydotblog reviewed (commented):
Working as expected.
